Port BZ 64097: fix services lookup

volosied opened the pull request:
Description: 

If META-INF/services/javax.el.ExpressionFactory does not contain a qualified class name in its first line (i.e. comments above the class name), then EL throws an exception. 

Fixes #18419 

Comments were not ignored with the previous lookup implementation. 


See original patch here: https://github.com/apache/tomcat/commit/31bce633bb5f90cf8d48074494e30254b939560d

volosied committed:
EL-3.0 API: Port BZ 64097: fix services lookup

volosied commented:
#build

volosied commented:
#build

pnicolucci reviewed (commented):
Vlad, can you add a test for this?

wraschke commented:
This is just for testing. It shouldn't interfere with any of your builds that you might have manually requested. I will cancel immediately.

#build

volosied committed:
Apply 64097 Fix in EL 2.2

volosied commented:
#build

volosied committed:
Create Service Lookup test under EL30MiscTests.java

Rename ServicesLookup test app

Add testEL22ServiceLookup test

Create jsp22Server.xml

Make EL30VarargsMethodMatchingTest run only in full

Create elMiscServer to avoid app complications

Apply formatting changes

Corrections

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.el.3.0_fat

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.el.3.0_fat

volosied commented on dev/com.ibm.ws.el.3.0_fat/fat/src/com/ibm/ws/el/fat/tests/EL30VarargsMethodMatchingTest.java:
```diff
@@ -20,12 +20,15 @@
 /**
  * Test EL 3.0 Method Matching when Varargs are used. See BZ 65358 
  */
+@Mode(TestMode.FULL)
```
This test should have been FULL, so I modified this test here.

volosied commented:
#build

volosied commented:
#build

pnicolucci commented on dev/com.ibm.ws.el.3.0_fat/fat/src/com/ibm/ws/el/fat/FATSuite.java:
```diff
@@ -21,6 +21,7 @@
 import com.ibm.ws.el.fat.tests.EL30LambdaExpressionsTest;
```
Can you update the copyright date range to include `2021`?

pnicolucci commented on dev/com.ibm.ws.el.3.0_fat/fat/src/com/ibm/ws/el/fat/tests/EL30MiscTests.java:
```diff
@@ -0,0 +1,140 @@
+        elServer.setMarkToEndOfLog();
+        elServer.restoreServerConfiguration();
+        // Wait for the application that is still deployed to start.
+        elServer.waitForConfigUpdateInLogUsingMark(Collections.singleton(ServiceLookup_AppName), true, "CWWKT0016I:.*ELServicesLookup.*");
```
Can we do this in a try-finally to ensure if there is an exception from one of the asserts that we actually restore the server.xml back to the original configuration?

volosied replied:
Let me know how it looks now

pnicolucci commented on dev/com.ibm.ws.el.3.0_fat/publish/files/serverConfigs/jsp22server.xml:
```diff
@@ -0,0 +1,21 @@
+        <feature>componenttest-1.0</feature>
+    </featureManager>
+
+</server>
```
Can you add a new line?

pnicolucci commented on dev/com.ibm.ws.el.3.0_fat/publish/servers/elMiscServer/bootstrap.properties:
```diff
@@ -0,0 +1,12 @@
+###############################################################################
+# Copyright (c) 2014 IBM Corporation and others.
+# All rights reserved. This program and the accompanying materials
```
should be `2021`

pnicolucci commented on dev/com.ibm.ws.el.3.0_fat/publish/files/serverConfigs/jsp22server.xml:
```diff
@@ -0,0 +1,21 @@
+        IBM Corporation - initial API and implementation
+ -->
+<server description="Server for testing EL 3.0 and EL 2.2 (through jsp-2.2) in a stand alone environment">
+
```
How is this actually used to test EL 3.0?

volosied replied:
Removed el-3.0 reference

pnicolucci commented on dev/com.ibm.ws.el.3.0_fat/test-applications/ELServicesLookup.war/resources/META-INF/services/javax.el.ExpressionFactory:
```diff
@@ -0,0 +1,2 @@
+# ExpressionFactory#getClassNameServices should ignore this comment -- OLGH18419 
+org.apache.el.ExpressionFactoryImpl
```
add new line

pnicolucci commented on dev/com.ibm.ws.el.3.0_fat/test-applications/ELServicesLookup.war/src/com/ibm/ws/el30/fat/servicelookup/TestServiceLookupServlet.java:
```diff
@@ -0,0 +1,34 @@
+      
+   }
+
+}
```
add new line

volosied commented:
#build

volosied commented:
#build

volosied commented:
#run-libby-bot

volosied commented:
Build failures related to GUI tests. Merging code in.
